### ISSUE TITLE
Issues with source-map generation

### ISSUE THREAD

**Issue by jbbr** [TypeStrong/ts-loader#81]
Enabling source map support in webpack config does not work properly with this loader. It depends on the `devtool` option choosen:

`devtool: 'source-map'`:
```
/home/user/Develop/myproject/node_modules/webpack-core/node_modules/source-map/lib/source-map/array-set.js:93
    throw new Error('No element indexed by ' + aIdx);
    ^

Error: No element indexed by 1
    at ArraySet_at [as at] (/home/user/Develop/myproject/node_modules/webpack-core/node_modules/source-map/lib/source-map/array-set.js:93:11)
    at SourceMapConsumer.<anonymous> (/home/user/Develop/myproject/node_modules/webpack-core/node_modules/source-map/lib/source-map/source-map-consumer.js:147:69)
    at Array.map (native)
    at SourceMapConsumer_eachMapping [as eachMapping] (/home/user/Develop/myproject/node_modules/webpack-core/node_modules/source-map/lib/source-map/source-map-consumer.js:146:16)
    at Function.SourceNode_fromStringWithSourceMap [as fromStringWithSourceMap] (/home/user/Develop/myproject/node_modules/webpack-core/node_modules/source-map/lib/source-map/source-node.js:84:26)
    at SourceMapSource.node (/home/user/Develop/myproject/node_modules/webpack-core/lib/SourceMapSource.js:41:20)
    at ReplaceSource.node (/home/user/Develop/myproject/node_modules/webpack-core/lib/ReplaceSource.js:55:29)
    at CachedSource.node (/home/user/Develop/myproject/node_modules/webpack-core/lib/CachedSource.js:41:22)
    at PrefixSource.node (/home/user/Develop/myproject/node_modules/webpack-core/lib/PrefixSource.js:27:26)
    at /home/user/Develop/myproject/node_modules/webpack-core/lib/ConcatSource.js:38:49
    at Array.map (native)
    at ConcatSource.node (/home/user/Develop/myproject/node_modules/webpack-core/lib/ConcatSource.js:37:60)
    at /home/user/Develop/myproject/node_modules/webpack-core/lib/ConcatSource.js:38:49
    at Array.map (native)
    at ConcatSource.node (/home/user/Develop/myproject/node_modules/webpack-core/lib/ConcatSource.js:37:60)
    at /home/user/Develop/myproject/node_modules/webpack-core/lib/ConcatSource.js:38:49
```

`devtool: 'eval-source-map'`:
```
ERROR in chunk main [entry]
client.bundle.js
No element indexed by 1
```

`devtool: 'eval'`: Everything works as expected (But source maps are not as readable as usually with the other options.

With https://github.com/s-panferov/awesome-typescript-loader I don't have this problem

**Comment by jbrantly** [comment]
Thanks for reporting this. I have a few questions for you.

1. Are you setting `sourceMap: true` in your tsconfig.json?
2. What version of TypeScript, ts-loader and webpack are you using?

[There is a test](https://github.com/TypeStrong/ts-loader/tree/master/test/sourceMaps) that has `devtool: 'source-map'` set so I'm curious what the difference is between that project and yours that might be causing the issue.

**Comment by jbbr** [comment]
@jbrantly thanks for the quick response.
I assume this is some kind of incompatibility with something in my webpack config. 
I'm using a webpack/Meteor skeleton as a base and added typescript loaders to it. Direct link to the webpack config directory with ts-loader enabled: https://github.com/jbbr/meteor-webpack-react/tree/13f05025a636c4b2677a818a45feb7834d52cc5d/webpack

Due to the meteor stuff it's a bit complicated... 

To answer your questions:

1. Yes
2. TypeScript 1.6.2, ts-loader 0.5.6, webpack 0.12.2

**Comment by jbrantly** [comment]
@jbbr I'm trying to duplicate this. I grabbed your repo and got the `ts-loader` branch and switched back to ts-loader. Running `node dev.js` doesn't result in any errors for me. Am I doing something wrong?

**Comment by jbrantly** [comment]
Actually, nvm, sorry. I had to change `eval` to `source-map`. I was able to duplicate the issue.

**Comment by jbrantly** [comment]
OK, so I've tracked this down to [how ts-loader is setting the `file` property](https://github.com/TypeStrong/ts-loader/blob/2f3faba02a7c3af9e9381db2716fc94080101fc1/index.ts#L539) on the sourcemap. Currently this code returns something like `path/to/loader/index.js!path/to/module.ts`. As near as I can tell this is technically correct since `file` is supposed to represent the ["name" of the output file](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit?hl=en_US&pli=1&pli=1#heading=h.qz3o9nc69um5). In this case it doesn't really have a name since it's not really a file. It's a file that is being transformed by a loader which is what `path/to/loader/index.js!path/to/module.ts` represents. In any case, something else later on in the webpack process which I haven't identified fails. If I remove `file` it fails. If I change to `path/to/module.js` it fails. If I change to `path/to/module.ts` then it works (which is really weird since that is the source module path. This is like assuming the output overwrites the source).

@sokra has mentioned that using [getCurrentRequest](https://github.com/webpack/less-loader/pull/33#issuecomment-71931194) or [this.request](https://github.com/webpack/webpack/issues/531#issuecomment-58921044) is the correct thing to do. It really does seem to me that `getCurrentRequest` is the right thing theoretically.

There is also a [bug](https://github.com/webpack/webpack/issues/1071) reported for webpack describing the same error but it's hard to tell if it's caused for the same reason.

In any case, I'm not 100% sure that this is a bug in ts-loader, I think it's more likely that it's a bug in webpack, BUT I'm willing to make this change so that this works. @blakeembrey Any thoughts?

**Comment by blakeembrey** [comment]
@jbrantly Honestly, not sure - I've only used the real filename for that property without issue. I'd probably have to jump into the Webpack source code to figure it out, but I don't have time to do that for a while.

**Comment by ShyykoSerhiy** [comment]
It looks like the root of issue is in babel loader after all.
https://github.com/babel/babel/blob/master/packages/babel/src/transformation/file/index.js#L472.
For some reason when tsx file has decorators inside of it corresponding sourceMap will have two sources and two sourcesContent. Guys from babel make some voodoo magic with replacing ```sources``` of a source map to ensure one source map in the end. That actually doesn't work when formal source map has two and more sources inside(which is true when decorators are involved).

Removing of this two lines:
```
mergedMap.sources = inputMap.sources;
mergedMap.file    = inputMap.file;
```
kind of fixes the problem, but probably guys from babel had reasons to put them there.

**Comment by jbbr** [comment]
wow, thanks for putting so much effort into research of this issue.

There seems to be lot's of issues on the web like this of people using the babel webpack loader. Suggests that it might be a babel bug.

Is there anything I can do to track it down? As I've already written, this error does not occur when using `awesome typescript loader`. Thus for me there is no hurry.
They must do something different then you or just added a workaround, I don't know the internals.

**Comment by jbrantly** [comment]
@jbbr `awesome-typescript-loader` is using `path/to/module.ts`, that's why it works. I'm OK with changing ts-loader to do the same even though I don't think the issue originates with ts-loader, at least until whatever downstream bug exists is fixed. It's easier for me to change ts-loader than it is to find and/or fix the downstream issue, at least right now :grin:

**Comment by ShyykoSerhiy** [comment]
Someone faster then me has already opened pull request to fix this issue in babel : https://github.com/babel/babel/pull/2522 .

**Comment by jbrantly** [comment]
I manually applied the fix in babel/babel#2522 and confirmed that it fixed this issue (or at least, it now builds without erroring. I did not check the validity of the sourcemap). I'm going to leave ts-loader code alone since I think it's technically doing the right thing.

**Comment by phestermcs** [comment]
Just FYI, I made the pull request in babel. @jbrantly I just added a comment there visualizing the mappings; should help you see that the fix actually works (that its not just building without errors), but that also Babel does a mediocre job making source maps to begin with.

In my case I'm using Intellij, which has the TS transpiler kinda built in so TS files are individually transpiled (and sourcemaps gen'd) by Intellij. So rather than using ts-loader, I'm using source-map-loader, then babel-loader.

I think there's an issue with webpack and/or source-map-loader, in how it changes the path of a file in a way inconsistent with how that file is named within an actual sourcemap. This resulted in the babel author overwriting (or rather restoring) the file name of the ts output before merging the maps, so that the sources entry in the map could be correctly located by the sourcemap merging logic. But it then also appears the author believed they had to overwrite the actual sourceMap.sources, and sourceMap.sourcesContent of the merged map as well, which was the root cause of the exception at the top of this issue; they did not have to. I only mention this in passing from reading ts-loader is naming the output file in a way that may somehow be related to understanding how source map merging is working in webpack.

**Comment by jbrantly** [comment]
@phestermcs Thanks for your work on that issue.

**Comment by sebald** [comment]
I encountered a similar issue. Here is the relevant part of my setup:

```javascript
module: {
    loaders: [

        // Support for Typescript
        { test: /\.ts$/, loader: 'ts' },

        // Support for Styles (CSS/Sass)
        {
            test:   /\.s(c|a)ss$/,
            loader: ExtractTextPlugin.extract(
                'style',
                `css?sourceMap!sass?sourceMap&includePaths[]=${bourbon.includePaths}`
            )
        }
    ]
},

plugins: [
    new ExtractTextPlugin('style.css', {
        allChunks: true
    })
]
```

With the `ExtraTextPlugin` entabled I get links to the TypeScript source like `test.scss:7`. Which is obviously not really the correct file! If I disable the plugin (via `disable: true`) everything works fine as long as I am using the `source-map` option.

Did anyone encounter a similar problem?

**Comment by jbrantly** [comment]
@sebald Thanks, I'll take a look. It was already on my list to see if this could be causing issues with other plugins like UglifyJS. It'll be interesting to see if it's the same root issue for ExtractTextPlugin.

**Comment by sebald** [comment]
@jbrantly Thx, for the insane quick reply. Let me know if I can help you with some more info. I also could try to "extract" the relevant part and upload it into a repo if you want.

I actually just checked the generated source and with the `ExtraTextPlugin` enabled, there is no TypeScript source map generted, which I find very confusing because Chrome has to get the information from somewhere o.O

**Comment by jbrantly** [comment]
So I decided to just make the change for this in ts-loader since I don't think it's harmful and the Babel PR doesn't seem to be moving very fast. This is published in v0.6.0.

@sebald I didn't get a chance to test your setup but if you could try again with v0.6.0 to see if it makes a difference.

**Comment by sebald** [comment]
@jbrantly sadly seems not to change anything :(

Loading `scss` files seems to fail with the update (maybe an error of sass-loader). Webpack runs w/o any errors but when I open the app in the browser webpack throws an `module not found` error.

Here is a link to a more complete [Webpack configuration](https://gist.github.com/sebald/a3889b598492bdcddc2b). Maybe it helps.

I would like to help out but this is my first Webpack setup (switched from Browserify). I will try to debug the source maps generation more thorough this week.

**Comment by ShyykoSerhiy** [comment]
I can confirm, that "Error: No element indexed by 1" is fixed in 0.6.0

**Comment by jbrantly** [comment]
@ShyykoSerhiy Thanks for the confirmation :+1:

@sebald Yea I wasn't really expecting it to :frowning: My guess is that your issue is probably unrelated to ts-loader but I can try to take a deeper look. Would you mind opening a new issue for it? Thanks!
